server/src/main/java/org/opensearch/cluster/routing/allocation/decider/DiskThresholdDecider.java

```diff
@@ -167,6 +174,42 @@ public static long sizeOfRelocatingShards(
     @Override
     public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
         ClusterInfo clusterInfo = allocation.clusterInfo();
+
+        /*
+         The following block enables allocation for remote shards within safeguard limits of the filecache.
+         */
+        if (REMOTE_CAPABLE.equals(getNodePool(node)) && REMOTE_CAPABLE.equals(getShardPool(shardRouting, allocation))) {
+            final Predicate<ShardRouting> shardRoutingPredicate = shard -> shard.primary()
+                && REMOTE_CAPABLE.equals(getShardPool(shard, allocation));
+            final List<ShardRouting> remoteShardsOnNode = shardsSatisfyingPredicate(node, shardRoutingPredicate);
+            final long currentNodeRemoteShardSize = remoteShardsOnNode.stream()
+                .map(ShardRouting::getExpectedShardSize)
+                .mapToLong(Long::longValue)
+                .sum();
+
+            final long shardSize = getExpectedShardSize(
+                shardRouting,
+                0L,
+                allocation.clusterInfo(),
+                allocation.snapshotShardSizeInfo(),
+                allocation.metadata(),
+                allocation.routingTable()
+            );
+
+            final FileCacheStats fileCacheStats = clusterInfo.getNodeFileCacheStats().getOrDefault(node.nodeId(), null);
+            final long nodeCacheSize = fileCacheStats != null ? fileCacheStats.getTotal().getBytes() : 0;
+            final long totalNodeRemoteShardSize = currentNodeRemoteShardSize + shardSize;
+
+            if (totalNodeRemoteShardSize > DATA_TO_FILE_CACHE_SIZE_RATIO * nodeCacheSize) {
+                return allocation.decision(
+                    Decision.NO,
+                    NAME,
+                    "file cache limit reached - remote shard size will exceed configured safeguard ratio"
+                );
+            }
+            return Decision.YES;
```

Comment:
@kotwanikunal I think it returns the "yes" decision prematurely, we still have to check the available free disk + watermarks otherwise even if we are good on cache sizes, we may not have enough disk space to keep segments.

Reply:
@reta - All of this is already taken care of. The `FsProbe` mechanism  was updated to consider the file cache capacity - https://github.com/opensearch-project/OpenSearch/blob/317dd03ed83110a5c4ab665f3682a22a7c1c864a/server/src/main/java/org/opensearch/monitor/fs/FsProbe.java#L81-L84

(The data for disk usage is fetched from `ClusterInfoService`, which instead calls the `NodeSerice`, which calls the `FsService` which uses `FSProbe` to get the available sizes. )